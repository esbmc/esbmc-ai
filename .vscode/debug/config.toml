ai_model = "openai:gpt-5-nano"
temp_auto_clean = true
#temp_file_dir = "temp"
allow_successful = false
loading_hints = false
source_code_format = "full"
dev_mode = true

[ai_custom."llama3.2:3b"]
server_type = "ollama"
url = "localhost:11434"
max_tokens = 100000

[ai_custom."llama3.3:latest"]
server_type = "ollama"
url = "localhost:11435"
max_tokens = 100000

[solution]
filenames = ["samples/threading.c"]

[verifier]
name = "esbmc"

[verifier.esbmc]
path = "~/.local/bin/esbmc"
params = [
    "--interval-analysis",
    "--memory-leak-check",
    "--goto-unwind",
    "--unlimited-goto-unwind",
    "--k-induction",
    "--state-hashing",
    "--add-symex-value-sets",
    "--k-step",
    "2",
    "--floatbv",
    "--unlimited-k-steps",
    "--context-bound",
    "2",
]
output_type = "full"
timeout = 60

[llm_requests]
max_tries = 5
timeout = 60
