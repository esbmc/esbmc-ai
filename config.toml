ai_model = "gpt-3.5-turbo"
temp_auto_clean = true
#temp_file_dir = "temp"
allow_successful = false
loading_hints = true
source_code_format = "full"

[verifier]
name = "esbmc"

[verifier.esbmc]
path = "~/.local/bin/esbmc"
params = [
    "--interval-analysis",
    "--goto-unwind",
    "--unlimited-goto-unwind",
    "--k-induction",
    "--state-hashing",
    "--add-symex-value-sets",
    "--k-step",
    "2",
    "--floatbv",
    "--unlimited-k-steps",
    "--compact-trace",
    "--context-bound",
    "2",
]
output_type = "full"
timeout = 60

[llm_requests]
max_tries = 5
timeout = 60

[user_chat]
temperature = 1.0

[fix_code]
temperature = 0.7
max_attempts = 5
message_history = "normal"

# PROMPT TEMPLATES - USER CHAT

[prompt_templates.user_chat]
initial = "Walk me through the source code, while also explaining the output of ESBMC at the relevant parts. You shall not start the reply with an acknowledgement message such as 'Certainly'."

[[prompt_templates.user_chat.system]]
role = "System"
content = "You are a security focused assistant that parses output from a program called ESBMC and explains the output to the user. ESBMC (the Efficient SMT-based Context-Bounded Model Checker) is a context-bounded model checker for verifying single and multithreaded C/C++, Kotlin, and Solidity programs. It can automatically verify both predefined safety properties (e.g., bounds check, pointer safety, overflow) and user-defined program assertions. You don't need to explain how ESBMC works, you only need to parse and explain the vulnerabilities that the output shows. For each line of code explained, say what the line number is as well. Do not answer any questions outside of these explicit parameters. If you understand, reply OK."

[[prompt_templates.user_chat.system]]
role = "AI"
content = "OK"

[[prompt_templates.user_chat.system]]
role = "System"
content = "Reply OK if you understand that the following text is the program source code:\n\n```c{source_code}```"

[[prompt_templates.user_chat.system]]
role = "AI"
content = "OK"

[[prompt_templates.user_chat.system]]
role = "System"
content = "Reply OK if you understand that the following text is the output from ESBMC:\n\n```\n{esbmc_output}\n```"

[[prompt_templates.user_chat.system]]
role = "AI"
content = "OK"

[[prompt_templates.user_chat.set_solution]]
role = "System"
content = "Here is the corrected code:\n\n```c\n{source_code_solution}```"

[[prompt_templates.user_chat.set_solution]]
role = "AI"
content = "OK"

# PROMPT TEMPLATES - FIX CODE

[prompt_templates.fix_code.base]
initial = "The ESBMC output is:\n\n```\n{esbmc_output}\n```\n\nThe source code is:\n\n```c\n{source_code}\n```\n Using the ESBMC output, show the fixed text."

[[prompt_templates.fix_code.base.system]]
role = "System"
content = "From now on, act as an Automated Code Repair Tool that repairs AI C code. You will be shown AI C code, along with ESBMC output. Pay close attention to the ESBMC output, which contains a stack trace along with the type of error that occurred and its location that you need to fix. Provide the repaired C code as output, as would an Automated Code Repair Tool. Aside from the corrected source code, do not output any other text."

[prompt_templates.fix_code."division by zero"]
initial = "The ESBMC output is:\n\n```\n{esbmc_output}\n```\n\nThe source code is:\n\n```c\n{source_code}\n```\n Using the ESBMC output, show the fixed text."

[[prompt_templates.fix_code."division by zero".system]]
role = "System"
content = "Here's a C program with a vulnerability:\n```c\n{source_code}\n```\nA Formal Verification tool identified a division by zero issue:\n{esbmc_output}\nTask: Modify the C code to safely handle scenarios where division by zero might occur. The solution should prevent undefined behavior or crashes due to division by zero. \nGuidelines: Focus on making essential changes only. Avoid adding or modifying comments, and ensure the changes are precise and minimal.\nGuidelines: Ensure the revised code avoids undefined behavior and handles division by zero cases effectively.\nGuidelines: Implement safeguards (like comparison) to prevent division by zero instead of using literal divisions like 1.0/0.0.Output: Provide the corrected, complete C code. The solution should compile and run error-free, addressing the division by zero vulnerability.\nStart the code snippet with ```c and end with ```. Reply OK if you understand."

[[prompt_templates.fix_code."division by zero".system]]
role = "AI"
content = "OK."
