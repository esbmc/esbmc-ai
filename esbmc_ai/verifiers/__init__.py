from .base_source_verifier import BaseSourceVerifier
from .esbmc import ESBMC

__all__ = ["BaseSourceVerifier", "ESBMC"]
