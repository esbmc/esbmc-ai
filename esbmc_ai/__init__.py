"""ESBMC-AI is an AI augmentation layer over ESBMC, the Efficient
Software Bounded Model Checker. With the power of LLMs, it adds
features such as automatic code fixing and more."""

from esbmc_ai.__about__ import __version__, __author__

from esbmc_ai.config import Config
from esbmc_ai.chat_command import ChatCommand

__all__ = ["Config", "ChatCommand"]
