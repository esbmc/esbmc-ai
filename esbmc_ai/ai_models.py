# Author: Yiannis Charalambous

from abc import abstractmethod
from typing import Any, Iterable, Optional, Union
from enum import Enum
from langchain_core.language_models import BaseChatModel
from pydantic.v1.types import SecretStr
from typing_extensions import override

from langchain_openai import ChatOpenAI
from langchain_ollama import ChatOllama

from langchain.prompts.chat import ChatPromptTemplate
from langchain.schema import (
    BaseMessage,
    PromptValue,
)


from esbmc_ai.api_key_collection import APIKeyCollection


class AIModel(object):
    """This base class represents an abstract AI model."""

    name: str
    tokens: int

    def __init__(
        self,
        name: str,
        tokens: int,
    ) -> None:
        self.name = name
        self.tokens = tokens

    @abstractmethod
    def create_llm(
        self,
        api_keys: APIKeyCollection,
        temperature: float = 1.0,
        requests_max_tries: int = 5,
        requests_timeout: float = 60,
    ) -> BaseChatModel:
        """Initializes a large language model model with the provided parameters."""
        raise NotImplementedError()

    @classmethod
    def convert_messages_to_tuples(
        cls, messages: Iterable[BaseMessage]
    ) -> list[tuple[str, str]]:
        """Converts messages into a format understood by the ChatPromptTemplate - since it won't format
        BaseMessage derived classes for some reason, but will for tuples, because they get converted into
        Templates in function `_convert_to_message`."""
        return [(message.type, str(message.content)) for message in messages]

    @classmethod
    def escape_messages(
        cls,
        messages: Iterable[BaseMessage],
        allowed_keys: list[str],
    ) -> Iterable[BaseMessage]:
        """Adds escape curly braces to the messages, will make sure that the sequential
        curly braces in the messages is even (and hence escaped). Will ignore curly braces
        with `allowed_keys`."""

        def add_safeguards(content: str, char: str, allowed_keys: list[str]) -> str:
            start_idx: int = 0
            while True:
                char_idx: int = content.find(char, start_idx)
                if char_idx == -1:
                    break
                # Count how many sequences of the char are occuring
                count: int = 1
                # FIXME Add bounds check here
                while (
                    len(content) > char_idx + count
                    and content[char_idx + count] == char
                ):
                    count += 1

                # Check if the next sequence is in the allowed keys, if it is, then
                # skip to the next one.
                is_allowed: bool = False
                for key in allowed_keys:
                    if key == content[char_idx + count : char_idx + count + len(key)]:
                        is_allowed = True
                        break

                # Now change start_idx to reflect new location. The start index is at the end of
                # all the chars (including the inserted one).
                start_idx = char_idx + count + 1

                # If inside allowed keys, then continue to next iteration.
                if is_allowed:
                    continue

                # Check if odd number (will need to add extra char)
                if count % 2 != 0:
                    content = (
                        content[: char_idx + count] + char + content[char_idx + count :]
                    )
            return content

        reversed_keys: list[str] = [key[::-1] for key in allowed_keys]

        result: list[BaseMessage] = []
        for msg in messages:
            content: str = str(msg.content)
            look_pointer: int = 0
            # Open curly check
            if content.find("{", look_pointer) != -1:
                content = add_safeguards(content, "{", allowed_keys)
            # Close curly check
            if content.find("}", look_pointer) != -1:
                # Do it in reverse with reverse keys.
                content = add_safeguards(content[::-1], "}", reversed_keys)[::-1]
            new_msg = msg.copy()
            new_msg.content = content
            result.append(new_msg)
        return result

    def apply_chat_template(
        self,
        messages: Iterable[BaseMessage],
        **format_values: Any,
    ) -> PromptValue:
        """Applies the formatted values onto the message chat template. For example,
        if the message contains the token {source}, then format_values contains a
        value for {source} then it will be substituted."""
        escaped_messages = AIModel.escape_messages(messages, list(format_values.keys()))
        message_tuples = AIModel.convert_messages_to_tuples(escaped_messages)
        return ChatPromptTemplate.from_messages(messages=message_tuples).format_prompt(
            **format_values,
        )


class AIModelOpenAI(AIModel):
    context_length_exceeded_error: str = "context_length_exceeded"
    """Error code for when the length has been reached."""

    @override
    def create_llm(
        self,
        api_keys: APIKeyCollection,
        temperature: float = 1.0,
        requests_max_tries: int = 5,
        requests_timeout: float = 60,
    ) -> BaseChatModel:
        assert api_keys.openai, "No OpenAI api key has been specified..."
        return ChatOpenAI(
            model=self.name,
            api_key=SecretStr(api_keys.openai),
            max_tokens=None,
            temperature=temperature,
            max_retries=requests_max_tries,
            timeout=requests_timeout,
            model_kwargs={},
        )

class OllamaAIModel(AIModel):
    def __init__(self, name: str, tokens: int, url: str) -> None:
        super().__init__(name, tokens)
        self.url: str = url
    
    @override
    def create_llm(self, api_keys: APIKeyCollection, temperature: float = 1, requests_max_tries: int = 5, requests_timeout: float = 60) -> BaseChatModel:
        # Ollama does not use API keys
        _ = api_keys
        _ = requests_max_tries
        return ChatOllama(
            base_url=self.url,
            model=self.name,
            temperature=temperature,
            client_kwargs={
                "timeout":requests_timeout,
            },
        )


class _AIModels(Enum):
    """Private enum that contains predefined AI Models. OpenAI models are not
    defined because they are fetched from the API."""

    # FALCON_7B = OllamaAIModel(...)
    pass


_custom_ai_models: list[AIModel] = []

_ai_model_names: set[str] = set(item.value.name for item in _AIModels)


def add_custom_ai_model(ai_model: AIModel) -> None:
    """Registers a custom AI model."""
    # Check if AI already already exists.
    if ai_model.name in _ai_model_names:
        raise Exception(f'AI Model "{ai_model.name}" already exists...')
    _ai_model_names.add(ai_model.name)
    _custom_ai_models.append(ai_model)


def is_valid_ai_model(
    ai_model: Union[str, AIModel], api_keys: Optional[APIKeyCollection] = None
) -> bool:
    """Accepts both the AIModel object and the name as parameter. It checks the
    openai servers to see if a model is defined on their servers, if not, then
    it checks the internally defined AI models list."""

    # Get the name of the model
    name: str = ai_model.name if isinstance(ai_model, AIModel) else ai_model

    # Try accessing openai api and checking if there is a model defined.
    # NOTE: This is not tested as no way to mock API currently.
    if api_keys and api_keys.openai:
        try:
            from openai import Client

            for model in Client(api_key=api_keys.openai).models.list().data:
                if model.id == name:
                    return True
        except ImportError:
            pass

    # Use the predefined list of models.
    return name in _ai_model_names


def _get_openai_model_max_tokens(name: str) -> int:
    """NOTE: OpenAI currently does not expose an API for getting the model
    length. Maybe add a config input value for this?"""

    # https://platform.openai.com/docs/models
    tokens = {
        "gpt-4o": 128000,
        "gpt-4": 8192,
        "gpt-3.5-turbo": 16385,
        "gpt-3.5-turbo-instruct": 4096,
    }

    # Split into - segments and remove each section from the end to find out
    # which one matches the most.

    # Base Case
    if name in tokens:
        return tokens[name]

    # Step Case
    name_split: list[str] = name.split("-")
    for i in range(1, name.count("-")):
        subname: str = "-".join(name_split[:-i])
        if subname in tokens:
            return tokens[subname]

    raise ValueError(f"Could not figure out max tokens for model: {name}")


def get_ai_model_by_name(
    name: str, api_keys: Optional[APIKeyCollection] = None
) -> AIModel:
    # Check OpenAI models.
    if api_keys and api_keys.openai:
        try:
            from openai import Client

            for model in Client(api_key=api_keys.openai).models.list().data:
                if model.id == name:
                    add_custom_ai_model(
                        AIModelOpenAI(
                            model.id,
                            _get_openai_model_max_tokens(model.id),
                        ),
                    )
        except ImportError:
            pass

    # Check AIModels enum.
    for enum_value in _AIModels:
        ai_model: AIModel = enum_value.value
        if name == ai_model.name:
            return ai_model

    # Check custom AI models.
    for custom_ai in _custom_ai_models:
        if name == custom_ai.name:
            return custom_ai

    raise Exception(f'The AI "{name}" was not found...')
