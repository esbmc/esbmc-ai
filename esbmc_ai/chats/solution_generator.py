# Author: Yiannis Charalambous 2023

"""Contains code for automatically repairing code using ESBMC."""

from typing import Optional
from langchain_core.language_models import BaseChatModel
from typing_extensions import override
from langchain.schema import BaseMessage

from esbmc_ai.chat_response import ChatResponse, FinishReason
from esbmc_ai.config import FixCodeScenario, default_scenario
from esbmc_ai.solution import SourceFile

from esbmc_ai.ai_models import AIModel
from esbmc_ai.verifiers.base_source_verifier import (
    BaseSourceVerifier,
    SourceCodeParseError,
)
from .base_chat_interface import BaseChatInterface


def get_source_code_formatted(
    verifier: BaseSourceVerifier,
    source_code_format: str,
    source_code: str,
    esbmc_output: str,
) -> str:
    """Gets the formatted output source code, based on the source_code_format
    passed."""
    match source_code_format:
        case "single":
            # Get source code error line from esbmc output
            line: Optional[int] = verifier.get_error_line_idx(esbmc_output)
            if line:
                return source_code.splitlines(True)[line]

            raise AssertionError(
                f"error line not found in esbmc output:\n{esbmc_output}"
            )
        case "full":
            return source_code
        case _:
            raise ValueError(
                f"Not a valid format for source code: {source_code_format}"
            )


class SolutionGenerator(BaseChatInterface):
    """SolutionGenerator is a simple conversation-based automated program repair
    class. The class works in a cycle, by first calling update_state with the
    new source_code and esbmc_output, then by calling generate_solution. The
    class supports scenarios to customize the system message and initial prompt
    based on the"""

    def __init__(
        self,
        scenarios: dict[str, FixCodeScenario],
        llm: BaseChatModel,
        ai_model: AIModel,
        verifier: BaseSourceVerifier,
        source_code_format: str = "full",
        esbmc_output_type: str = "full",
    ) -> None:
        """Initializes the solution generator."""

        super().__init__(
            ai_model=ai_model,
            llm=llm,
            system_messages=[],  # Empty as it will be updated in the update method.
        )

        self.scenarios: dict[str, FixCodeScenario] = scenarios
        self.scenario: Optional[str] = None

        self.verifier: BaseSourceVerifier = verifier

        self.esbmc_output_type: str = esbmc_output_type
        self.source_code_format: str = source_code_format

        self.source_code_raw: Optional[str] = None
        self.source_code_formatted: Optional[str] = None
        self.esbmc_output: Optional[str] = None
        self.invokations: int = 0

    @override
    def compress_message_stack(self) -> None:
        # Resets the conversation - cannot summarize code
        # If generate_solution is called after this point, it will start new
        # with the currently set state.
        self.messages: list[BaseMessage] = []
        self.invokations = 0

    @staticmethod
    def extract_code_from_solution(solution: str) -> str:
        """Strip the source code of any leftover text as sometimes the AI model
        will generate text and formatting despite being told not to."""
        try:
            code_start: int = solution.index("```") + 3
            assert code_start != -1

            # Remove up until the new line, because usually there's a language
            # specification after the 3 ticks ```c...
            code_start = solution.index("\n", code_start) + 1
            assert code_start != -1

            code_end: int = solution[::-1].index("```")
            assert code_start != -1

            # -4 = 3 backticks and also the \n before the backticks.
            code_end: int = len(solution) - 4 - code_end
            # +1 because of edge cases as in test_get_code_from_solution
            assert code_start <= code_end + 1

            solution = solution[code_start:code_end]
        except (ValueError, AssertionError):
            pass
        return solution

    def update_state(self, source_code: str, esbmc_output: str) -> None:
        """Updates the latest state of the code and ESBMC output. It also updates
        the scenario, which is the type of error that ESBMC has shown. This should be
        called before generate_solution."""

        self.scenario = self.verifier.get_error_scenario(esbmc_output)
        self.source_code_raw = source_code

        # Format ESBMC output
        try:
            self.esbmc_output = self.verifier.apply_formatting(
                verifier_output=esbmc_output,
                format=self.esbmc_output_type,
            )
        except SourceCodeParseError:
            # When clang output is displayed, show it entirely as it doesn't get very
            # big.
            self.esbmc_output = esbmc_output

        # Format source code
        self.source_code_formatted = get_source_code_formatted(
            verifier=self.verifier,
            source_code_format=self.source_code_format,
            source_code=source_code,
            esbmc_output=self.esbmc_output,
        )

    def _get_system_messages(
        self, override_scenario: Optional[str] = None
    ) -> tuple[BaseMessage, ...]:
        if override_scenario:
            system_messages = self.scenarios[override_scenario].system
        else:
            assert self.scenario, "Call update or set the scenario"
            if self.scenario in self.scenarios:
                system_messages = self.scenarios[self.scenario].system
            else:
                system_messages = self.scenarios[default_scenario].system

        assert isinstance(system_messages, tuple)
        assert all(isinstance(msg, BaseMessage) for msg in system_messages)
        return system_messages

    def _get_initial_message(
        self, override_scenario: Optional[str] = None
    ) -> BaseMessage:
        if override_scenario:
            return self.scenarios[override_scenario].initial
        else:
            assert self.scenario, "Call update or set the scenario"
            if self.scenario in self.scenarios:
                return self.scenarios[self.scenario].initial
            else:
                return self.scenarios[default_scenario].initial

    def generate_solution(
        self,
        override_scenario: Optional[str] = None,
        ignore_system_message: bool = False,
    ) -> tuple[str, FinishReason]:
        """Prompts the LLM to repair the source code using the verifier output.
        If this is the first time the method is called, the system message will
        be sent to the LLM, unless ignore_system_message is True. Then the
        initial prompt will be sent.

        In subsequent invokations of generate_solution, the initial prompt will
        be used only.

        So the system messages and initial message should each include at least
        {source_code} and {esbmc_output} so that they are substituted into the
        message.

        Queries the AI model to get a solution. Accepts an override scenario
        parameter, in which case the scenario won't be resolved automatically."""

        assert (
            self.source_code_raw is not None
            and self.source_code_formatted is not None
            and self.esbmc_output is not None
            and self.scenario is not None
        ), "Call update_state before calling generate_solution."

        # Show system message
        if not ignore_system_message and self.invokations <= 0:
            # Get scenario system messages and push it to message stack. Don't
            # push to system message stack because we want to regenerate from
            # the beginning at every reset.
            system_messages: tuple[BaseMessage, ...] = self._get_system_messages(
                override_scenario=override_scenario
            )
            if len(system_messages) > 0:
                self.push_to_message_stack(system_messages)

        # Get scenario initial message and push it to message stack
        self.push_to_message_stack(
            self._get_initial_message(override_scenario=override_scenario)
        )

        self.invokations += 1

        error_type: Optional[str] = self.verifier.get_error_type(self.esbmc_output)

        # Apply template substitution to message stack
        self.apply_template_value(
            **self.get_canonical_template_keys(
                source_code=self.source_code_formatted,
                esbmc_output=self.esbmc_output,
                error_line=str(self.verifier.get_error_line(self.esbmc_output)),
                error_type=error_type if error_type else "unknown error",
            )
        )

        # Generate the solution
        response: ChatResponse = self.send_message()
        solution: str = str(response.message.content)

        solution = SolutionGenerator.extract_code_from_solution(solution)

        # Post process source code
        # If source code passed to LLM is formatted then we need to recombine to
        # full source code before giving to ESBMC
        match self.source_code_format:
            case "single":
                # Get source code error line from esbmc output
                line: Optional[int] = self.verifier.get_error_line_idx(
                    self.esbmc_output
                )
                assert line, (
                    "fix code command: error line could not be found to apply "
                    "brutal patch replacement"
                )
                solution = SourceFile.apply_line_patch(
                    self.source_code_raw, solution, line, line
                )

        return solution, response.finish_reason
