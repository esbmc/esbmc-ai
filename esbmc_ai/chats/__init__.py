# Author: Yiannis Charalambous

"""This module contains different chat interfaces. Along with `BaseChatInterface`
that provides necessary boilet-plate for implementing an LLM based chat."""

from .base_chat_interface import BaseChatInterface
from .latest_state_solution_generator import LatestStateSolutionGenerator
from .solution_generator import SolutionGenerator
from .user_chat import UserChat

__all__ = [
    "BaseChatInterface",
    "LatestStateSolutionGenerator",
    "SolutionGenerator",
    "UserChat",
]
