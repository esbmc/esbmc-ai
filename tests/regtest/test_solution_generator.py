# Author: Yiannis Charalambous


from langchain.schema import HumanMessage, SystemMessage
from langchain_core.language_models import FakeListChatModel

from esbmc_ai.ai_models import AIModel
from esbmc_ai.verifiers.dummy_verifier import DummyVerifier
from esbmc_ai.chats.solution_generator import SolutionGenerator
from esbmc_ai.config import FixCodeScenario


def test_generate_solution(regtest) -> None:
    with open(
        "tests/samples/esbmc_output/line_test/cartpole_95_safe.c-amalgamation-80.c", "r"
    ) as file:
        esbmc_output: str = file.read()

    verifier = DummyVerifier(responses=[esbmc_output] * 2)

    chat = SolutionGenerator(
        scenarios={
            "base": FixCodeScenario(
                initial=HumanMessage(
                    "{source_code}{esbmc_output}{error_line}{error_type}"
                ),
                system=(
                    SystemMessage(
                        content="System:{source_code}{esbmc_output}{error_line}{error_type}"
                    ),
                ),
            )
        },
        verifier=verifier,
        ai_model=AIModel("test", 10000000),
        llm=FakeListChatModel(responses=["22222", "33333"]),
        source_code_format="full",
        esbmc_output_type="full",
    )

    chat.update_state("11111", esbmc_output)
    sol, res = chat.generate_solution(ignore_system_message=False)

    with regtest:
        print(res)
        print(sol)

    chat.update_state("11111", esbmc_output)
    sol, res = chat.generate_solution(ignore_system_message=False)

    with regtest:
        print(res)
        print(sol)
