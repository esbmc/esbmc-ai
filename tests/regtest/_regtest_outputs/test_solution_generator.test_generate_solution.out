FinishReason.stop
22222
FinishReason.stop
33333
